doc/designs/fips_indicator.md

```diff
@@ -0,0 +1,285 @@
+- Documentation is required to demonstrate how to use indicators for each approved cryptographic algorithm.
+- Testing is required to execute all services and verify that the indicator provides an unambiguous indication of whether the service utilizes an approved cryptographic algorithm, security function or process in an approved manner or not.
+- The Security Policy may require updates related to indicators. AWS/google have added a table in their security policy called ‘Non-Approved Algorithms not allowed in the approved mode of operation’. An example is RSA with a keysize of < 2048 bits (which has been enforced by [4]).
+
+Legacy Support
+--------------
+
+Due to key transitions [4] we may have some legacy algorithms that are in a state of only being approved for processing (verification, decryption, validation), and not for protection (signing, encrypting, keygen).
+For example DSA.
+
+The options are:
+
+- Completely remove the algorithm from the FIPS provider. This is simple but means older applications can no longer process existing data, which is not ideal.
+- Allow the algorithm but make it not approved with an context specific indicator.
+
+It is safer to make the protection operations fail rather than use an indicator.
+The processing operation for DSA would set the indicator to approved.
+
+Security Checks
+---------------
+
+OpenSSL currently defines configurable FIPS options.
+These options are supplied via the FIPS configuration file - which is normally setup via fipsinstall.
+
+- FIPS_FEATURE_CHECK(FIPS_security_check_enabled, fips_security_checks, 1)
+- FIPS_FEATURE_CHECK(FIPS_tls_prf_ems_check, fips_tls1_prf_ems_check, 0)
+- FIPS_FEATURE_CHECK(FIPS_restricted_drbg_digests_enabled, 0)
+- OSSL_PROV_FIPS_PARAM_CONDITIONAL_ERRORS selftest_params.conditional_error_check
+
+The following functions are available in providers/common/security_check.c.
+
+- ossl_rsa_check_key()
+- ossl_ec_check_key()
+- ossl_dsa_check_key()
+- ossl_dh_check_key()
+- ossl_digest_get_approved_nid_with_sha1()
+- ossl_digest_is_allowed()
+
+Anywhere where these functions are called an indicator MAY be required.
+Because these options are available I do not think it is sufficient to
+document this in the security policy.
+
+Each of these functions contains code of the following form:
+
+``` c
+#if !defined(OPENSSL_NO_FIPS_SECURITYCHECKS)
+    if (ossl_securitycheck_enabled(ctx)) {
+      // Do some checks, and maybe return 0 for a hard failure
+      ...
+    }
+}
+```
+
+OPENSSL_NO_FIPS_SECURITYCHECKS is also a configurable option
+If the security checks are not enabled then it is unapproved?
+
+Implementation options
+----------------------
+
+The above requirements indicate 2 options.
```

Comment:
Suggest adding the third option mooted by OTC yesterday (which amounts to _implicit indicators_):

- Detect non-FIPS operations and return an error.
  - For situations where we cannot know, err generously seems sensible (but up for discussion).
- Have a flag param to stop this behaviour to permit legacy access.
  - The flag could be either global (from `fipsmodule.cnf`) or per algorithm context.
    - A global flag is easier but dangerous.
    - A per context flag makes it obvious when the caller is possibly bypassing FIPS standards (because setparam).
    - A per context flag is **not** an indicator, it only indicates that usage **might not** be FIPS approved.
    - IMO per context is the better option.
  - "legacy" isn't a suitable name for this param IMO.
  - The flag could replace the ad-hoc settings in `fipsmodule.cnf` (but backward compatibility).
    - Nonetheless, it could replace all ad-hoc settings moving forwards.
* The complexity lies between the two other options because we still need to instrument all FIPS mandated checks but we don't need to add an indicator param everywhere.
* Non-FIPS algorithms are still possible either because they aren't FIPS approved (e.g. 3DES wouldn't need the flag param) or the flag is deliberately set.
* The default and other providers will ignore this flag param, so the solution won't break them.

Reply:
I must admit I don't fully grasp the reasons behind this approach. If someone is going through the trouble of running a validated version of the FIPS provider, do they even have the liberty of setting these flags to the "non-compliant" option? Would the security policy not mandate that these flags are always set to the "compliant" options? It's my understanding that compliance is black and white: either the module is compliant, or it is not.

But perhaps the various programs that use FIPS (e.g. FedRAMP) have a different view.

Reply:
If it is just black and white we wouldnt need indicators right?

Reply:
I think that's a little blunt. Indicators would indicate a non-approved service, which is explicitly within the scope of FIPS. E.g. there are requirements specifically for non-approved services, they are tested and listed in the security policy. They basically shift the responsibility to the operator who is supposed to check the indicator and then do... something.

Reply:
No, @slontis is spot on.  FIPS has a number of grey areas and permits backward compatibility that isn't within current guidelines.

FIPS allows old algorithms to be used in legacy situations -- e.g. triple DES was (until recently) allowed for decryption but not encryption.

Similarly for public key operations with short key sizes.  Likewise for MACs with short keys.  OpenSSL often (but not always) cannot distinguish between an operation to support something legacy from an operation that must be strictly FIPS approved as per the validated standard.  I.e. there **must** be a way for the caller to distinguish the two.

- Indicators are one approach: a caller can ask if what I just did was within the letter of the standards or not.
- The approach suggested by the OTC is the reverse: the caller has to intentionally say that the following might not meet the standards but allow it nonetheless.

Reply:
The only difference between setting the flag to allow non-approved and the explicit indicators (unless we implement both) is that without an explicit indicator setting the flag automatically means the operation is non-approved even though it was actually an approved operation.

On the other hand having to set such flag for non-approved operation makes it much more clean for applications to keep the FIPS compliance. Otherwise they would have to add the indicator check to each and every call into OpenSSL. And how is an application supposed to do it when it for example uses the FIPS module indirectly, i.e. via libssl or some third party library?

Reply:
No, setting the non-approved flag doesn't make any subsequent operation automatically non-approved.  So long as only approved algorithms are used in accordance with the standards, after setting the non-approved flag, the operation will still be approved.  It's up to the user to adhere if they set the non-approved flag.  This needs to be documented.

Explicit indicators are also quite a bit more effort, every algorithm beyond those in doubt needs to be instrumented to return "approved".

Agreed about the flag making compliance more obvious.  If you want non-FIPS, you **must** set the flag explicitly, if you don't whatever you do will either be approved or will error.  Checking the indicator after the operations is a far larger onus on the application IMO and one that will be ignored.

Upgrading the FIPS provider will cause applications to fail due to the changing standards.  That's part of FIPS compliance and application/library writers will have to deal with it.  While I agree that we should provide a work around, I'm far less sure that it should be enabled by default.

Reply:
So these flags would be set programmatically by the caller, rather than in the config file? Would the default behavior (i.e. the flag is not set) be compliant?

Reply:
Compliant by default would be my expectation, however the OTC might decide compatible by default instead.